src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -4125,6 +4177,161 @@ private SqlFedAuthToken getFedAuthToken(SqlFedAuthInfo fedAuthInfo) throws SQLSe
+            throw new SQLServerException(form.format(new Object[] {authenticationString}), null, 0, null);
+        }
+    }
+
+    private SqlFedAuthToken getMSIAuthToken(String resource, String msiClientId) throws SQLServerException {
+        // IMDS upgrade time can take up to 70s
+        final int imdsUpgradeTimeInMs = 70 * 1000;
+        final List<Integer> retrySlots = new ArrayList<>();
+        final String msiEndpoint = System.getenv("MSI_ENDPOINT");
+        final String msiSecret = System.getenv("MSI_SECRET");
+
+        StringBuilder urlString = new StringBuilder();
+        int retry = 1, maxRetry = 1;
+
+        boolean isAzureFunction = null != msiEndpoint && !msiEndpoint.isEmpty() && null != msiSecret
+                && !msiSecret.isEmpty();
+
+        if (isAzureFunction) {
+            urlString.append(msiEndpoint).append("?api-version=2017-09-01&resource=").append(resource);
+        } else {
+            urlString.append(ActiveDirectoryAuthentication.AZURE_REST_MSI_URL).append("&resource=").append(resource);
+            // Retry acquiring access token upto 20 times due to possible IMDS upgrade (Applies to VM only)
+            maxRetry = 20;
+            // Simplified variant of Exponential BackOff
+            for (int x = 0; x < maxRetry; x++) {
+                retrySlots.add(500 * ((2 << 1) - 1) / 1000);
+            }
+        }
+
+        // Append Client Id if available
+        if (null != msiClientId && !msiClientId.isEmpty()) {
+            if (isAzureFunction) {
+                urlString.append("&clientid=").append(msiClientId);
+            } else {
+                urlString.append("&client_id=").append(msiClientId);
+            }
+        }
+
+        // Loop while maxRetry reaches its limit
+        while (retry <= maxRetry) {
+            HttpURLConnection connection = null;
+
+            try {
+                connection = (HttpURLConnection) new URL(urlString.toString()).openConnection();
+                connection.setRequestMethod("GET");
+
+                if (isAzureFunction) {
+                    connection.setRequestProperty("Secret", msiSecret);
+                    if (connectionlogger.isLoggable(Level.FINER)) {
+                        connectionlogger.finer(toString() + " Using Azure Function/App Service MSI auth: " + urlString);
+                    }
+                } else {
+                    connection.setRequestProperty("Metadata", "true");
+                    if (connectionlogger.isLoggable(Level.FINER)) {
+                        connectionlogger.finer(toString() + " Using Azure MSI auth: " + urlString);
+                    }
+                }
+
+                connection.connect();
+
+                try (InputStream stream = connection.getInputStream()) {
+
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(stream, UTF_8), 100);
+                    String result = reader.readLine();
```

Comment:
Is it guaranteed that the response is only one line?

Reply:
Yes.